### PR TITLE
CFE-3521: Added checks to return value from getpwuid & getgrgid

### PR DESCRIPTION
**Pull request by larsewi**
Changelog: Title
Ticket: CFE-3521
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

### CODE REVIEW

File: cf-agent/simulate_mode.c [hunk 1]
```diff
@@ -63,8 +63,35 @@ static inline void PrintDelimiter()
...
+        }
+        return;
+    }
+
     struct group *group = getgrgid(st->st_gid);
+    if (group == NULL)
+    {
+        if (errno == 0)
+        {
+            Log(LOG_LEVEL_ERR, "Failed to get file group name: non-existent group id '%d'", st->st_gid);
```
Comment (larsewi) [inline cf-agent/simulate_mode.c:87]: `%d` should be changed to `%ju` ?

Comment (olehermanse) [reply cf-agent/simulate_mode.c:87]: What is the type of `st->st_gid`?

Comment (larsewi) [reply cf-agent/simulate_mode.c:87]: unsigned int

File: cf-agent/simulate_mode.c [hunk 2]
```diff
@@ -63,8 +63,35 @@ static inline void PrintDelimiter()
...
+    {
+        if (errno == 0)
+        {
+            Log(LOG_LEVEL_ERR, "Failed to get file owner name: non-existent user id %d", st->st_uid);
+        }
+        else 
+        {
+            Log(LOG_LEVEL_ERR, "Failed to get file owner name: %s", GetErrorStr());
+        }
+        return;
```
Comment (vpodzime) [inline cf-agent/simulate_mode.c:79]: I'd prefer if only the UID/GID-related bits were  omitted in the output. We can still report everything else from the stat info.

File: cf-agent/simulate_mode.c (hunk 1 above)
Comment (vpodzime) [reply cf-agent/simulate_mode.c:87]: Yes, `%ju` and `(uintmax_t)st->st_gid` would be better.

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise.

File: cf-agent/simulate_mode.c (hunk 2 above)
Comment (vpodzime) [reply cf-agent/simulate_mode.c:79]: I don't see how was my comment resolved in the code changes. :wink:

Comment (larsewi) [reply cf-agent/simulate_mode.c:79]: Sry @vpodzime , I must have misunderstood. I removed UID/GID from error output, did you mean to omit it in stdout output? Could you please elaborate?

Comment (olehermanse) [reply cf-agent/simulate_mode.c:79]: Maybe confusion between omit / emit ? Opposites ;)

**Comment by larsewi** [comment]
@cf-bottom jenkins please!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6004/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6004/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6004/

**Review by vpodzime — APPROVED** [review]
Looks good to me now.

File: cf-agent/simulate_mode.c [hunk 3]
```diff
@@ -87,10 +85,47 @@ static void ManifestStatInfo(const struct stat *st)
...
            (uintmax_t)st->st_uid, owner->pw_name);
-    printf("Gid: (%ju/%s)\n",
-           (uintmax_t)st->st_gid, group->gr_name);
+    }
+    else
+    {
+        if (errno == 0)
+        {
+            Log(LOG_LEVEL_ERR, "Failed to get file owner name: non-existent user id '%ju'", 
+                (uintmax_t)st->st_uid);
```
Comment (olehermanse) [inline cf-agent/simulate_mode.c:101]: Here and in other places, our style guide says there should be a space after the type cast.

https://github.com/cfengine/core/blob/master/CONTRIBUTING.md#formatting--whitespace

File: cf-agent/simulate_mode.c [hunk 4]
```diff
@@ -87,10 +85,47 @@ static void ManifestStatInfo(const struct stat *st)
...
+            Log(LOG_LEVEL_ERR, "Failed to get file owner name: %s", 
+                GetErrorStr());
+        }
+    }
+        
+    errno = 0;
+    struct group *group = getgrgid(st->st_gid);
+    if (group != NULL) {
+        printf("Gid: (%ju/%s)\n",
+            (uintmax_t)st->st_gid, group->gr_name);
```
Comment (olehermanse) [inline cf-agent/simulate_mode.c:114]: Alignment of arguments is off here, I would just put it on one line.

**Review by olehermanse — APPROVED** [review]

**Review by vpodzime — APPROVED** [review]
ACK

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please :)

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6010/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6010/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6010/

**Pull request merged**